sklearn/neighbors/_dist_metrics.pyx

```diff
@@ -689,13 +670,13 @@ cdef class MahalanobisDistance(DistanceMetric):
 
         # compute (x1 - x2).T * VI * (x1 - x2)
         for i in range(size):
-            self.vec_ptr[i] = x1[i] - x2[i]
+            self.vec[i] = x1[i] - x2[i]
 
         for i in range(size):
             tmp = 0
             for j in range(size):
-                tmp += self.mat_ptr[i * size + j] * self.vec_ptr[j]
-            d += tmp * self.vec_ptr[i]
+                tmp += self.mat[i, j] * self.vec[j]
```

Comment:
Could it be the cause of the performance regression?

It think it would be worth re-running the benchmarks another time to check whether the 1.5x slowdown is really reproducible for this one.

Reply:
The

```
#cython: boundscheck=False
#cython: wraparound=False
#cython: cdivision=True
```

should make it such that that numpy-style indexing should be as fast as pointer-style indexing (unless I missed something).

Reply:
I agree.

To me, the previous results can be explained by a change of the machine load (`EuclideanDistance` really has no reason to have better performance on this branch).

Reply:
Can you just re-run the benchmarks once more to confirm this?

Reply:
I started them again this morning, they will be out soon.

To me there must be no changes.

Reply:
Here are the results.

```
       before           after         ratio
     [b5e5db4a]       [47438839]
     <main>           <remove-old-np-compat-fixtures>
+         864±2μs      1.35±0.01ms     1.56   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(100, 10)
+      86.4±0.6ms        134±0.7ms     1.55   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(1000, 10)
+      8.99±0.01s       13.9±0.01s     1.54   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(10000, 10)
+      28.5±0.1ms       42.4±0.2ms     1.49   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(1000, 5)
+       291±0.8μs        431±0.9μs     1.48   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(100, 5)
+      3.21±0.01s       4.57±0.01s     1.42   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(10000, 5)
+      8.33±0.02s       10.9±0.03s     1.31   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(1000, 100)
+      85.2±0.1ms        112±0.3ms     1.31   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(100, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

There's solely a significant slow down for `MahalanobisDistance`.  
This must be linked to the patch there but according to the directive definitions as you pointer out, this _should not_ happen. :thinking: 

Does this relate in some ways to https://github.com/cython/cython/issues/2987? I am searching for similar issues.

<details>
<summary> Full report </summary>
<pre>
 → asv continuous -b Distance main remove-old-np-compat-fixtures
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl.
·· Building 47438839 <remove-old-np-compat-fixtures> for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl.............................................................
·· Installing 47438839 <remove-old-np-compat-fixtures> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl.
· Running 14 total benchmarks (2 commits * 1 environments * 7 benchmarks)
[  0.00%] · For scikit-learn commit 47438839 <remove-old-np-compat-fixtures> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[  7.14%] ···  distance_metric.ChebyshevDistanceBenchmark.time_pairwise                                                                        ok
[  7.14%] ··· ======= ============= ============= ============
              --                         d                    
              ------- ----------------------------------------
                 n          5             10          100     
              ======= ============= ============= ============
                100     81.9±0.4μs    133±0.7μs    949±0.7μs  
                1000   7.88±0.06ms   13.0±0.09ms   94.8±0.7ms 
               10000    1.07±0.01s    1.58±0.01s   10.7±0.01s 
              ======= ============= ============= ============

[ 14.29%] ···  distance_metric.EuclideanDistanceBenchmark.time_pairwise                                                                        ok
[ 14.29%] ··· ======= ============= ============= ============
              --                         d                    
              ------- ----------------------------------------
                 n          5             10          100     
              ======= ============= ============= ============
                100     79.0±0.4μs    114±0.6μs    979±0.5μs  
                1000   7.44±0.04ms   10.9±0.06ms   97.9±0.1ms 
               10000     1.05±0s       1.41±0s     10.8±0.01s 
              ======= ============= ============= ============

[ 21.43%] ···  distance_metric.MahalanobisDistanceBenchmark.time_pairwise                                                              1/9 failed
[ 21.43%] ··· ======= ============ ============= ============
              --                         d                   
              ------- ---------------------------------------
                 n         5             10          100     
              ======= ============ ============= ============
                100    431±0.9μs    1.35±0.01ms   112±0.3ms  
                1000   42.4±0.2ms    134±0.7ms    10.9±0.03s 
               10000   4.57±0.01s    13.9±0.01s     failed   
              ======= ============ ============= ============

[ 28.57%] ···  distance_metric.ManhattanDistanceBenchmark.time_pairwise                                                                        ok
[ 28.57%] ··· ======= ============= ============= ============
              --                         d                    
              ------- ----------------------------------------
                 n          5             10          100     
              ======= ============= ============= ============
                100     78.1±0.2μs    112±0.3μs     939±1μs   
                1000   7.38±0.02ms   10.8±0.05ms   93.8±0.2ms 
               10000     1.04±0s       1.38±0s     10.4±0.01s 
              ======= ============= ============= ============

[ 35.71%] ···  distance_metric.MinkowskiDistanceBenchmark.time_pairwise                                                                1/9 failed
[ 35.71%] ··· ======= ============ ============= =============
              --                         d                    
              ------- ----------------------------------------
                 n         5             10           100     
              ======= ============ ============= =============
                100     1.59±0ms    2.74±0.01ms   24.6±0.05ms 
                1000   159±0.7ms      276±2ms      2.42±0.01s 
               10000   16.2±0.02s    28.4±0.1s       failed   
              ======= ============ ============= =============

[ 42.86%] ···  distance_metric.SEuclideanDistanceBenchmark.time_pairwise                                                                   failed
[ 42.86%] ··· ======= ======== ======== ========
              --                  d             
              ------- --------------------------
                 n       5        10      100   
              ======= ======== ======== ========
                100    failed   failed   failed 
                1000   failed   failed   failed 
               10000   failed   failed   failed 
              ======= ======== ======== ========

[ 50.00%] ···  distance_metric.WMinkowskiDistanceBenchmark.time_pairwise                                                               1/9 failed
[ 50.00%] ··· ======= ============= ============= =============
              --                          d                    
              ------- -----------------------------------------
                 n          5             10           100     
              ======= ============= ============= =============
                100    1.58±0.01ms   2.80±0.01ms   24.5±0.05ms 
                1000    160±0.2ms      278±1ms      2.39±0.01s 
               10000     16.4±0s      28.3±0.02s      failed   
              ======= ============= ============= =============

[ 50.00%] · For scikit-learn commit b5e5db4a <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl..
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 57.14%] ···  distance_metric.ChebyshevDistanceBenchmark.time_pairwise                                                                        ok
[ 57.14%] ··· ======= ============= ============= ============
              --                         d                    
              ------- ----------------------------------------
                 n          5             10          100     
              ======= ============= ============= ============
                100     81.5±0.3μs    133±0.5μs    945±0.8μs  
                1000   7.72±0.03ms   12.9±0.06ms   94.7±0.2ms 
               10000    1.08±0.01s    1.58±0.01s   10.7±0.01s 
              ======= ============= ============= ============

[ 64.29%] ···  distance_metric.EuclideanDistanceBenchmark.time_pairwise                                                                        ok
[ 64.29%] ··· ======= ============= ============= ============
              --                         d                    
              ------- ----------------------------------------
                 n          5             10          100     
              ======= ============= ============= ============
                100     78.9±0.2μs    113±0.6μs     977±2μs   
                1000   7.38±0.03ms   10.9±0.03ms   97.8±0.2ms 
               10000    1.05±0.01s    1.40±0.01s   10.8±0.02s 
              ======= ============= ============= ============

[ 71.43%] ···  distance_metric.MahalanobisDistanceBenchmark.time_pairwise                                                              1/9 failed
[ 71.43%] ··· ======= ============ ============ ============
              --                        d                   
              ------- --------------------------------------
                 n         5            10          100     
              ======= ============ ============ ============
                100    291±0.8μs     864±2μs     85.2±0.1ms 
                1000   28.5±0.1ms   86.4±0.6ms   8.33±0.02s 
               10000   3.21±0.01s   8.99±0.01s     failed   
              ======= ============ ============ ============

[ 78.57%] ···  distance_metric.ManhattanDistanceBenchmark.time_pairwise                                                                        ok
[ 78.57%] ··· ======= ============= ============= ============
              --                         d                    
              ------- ----------------------------------------
                 n          5             10          100     
              ======= ============= ============= ============
                100     81.4±0.8μs    116±0.9μs    941±0.9μs  
                1000   7.71±0.05ms   11.1±0.05ms    94.0±1ms  
               10000     1.09±0s       1.44±0s     10.4±0.01s 
              ======= ============= ============= ============

[ 85.71%] ···  distance_metric.MinkowskiDistanceBenchmark.time_pairwise                                                                  1/9 failed
[ 85.71%] ··· ======= ============= ============= =============
              --                          d                    
              ------- -----------------------------------------
                 n          5             10           100     
              ======= ============= ============= =============
                100    1.59±0.01ms   2.78±0.01ms   24.6±0.04ms 
                1000     159±1ms       277±2ms      2.42±0.01s 
               10000    16.2±0.02s    28.2±0.3s       failed   
              ======= ============= ============= =============

[ 92.86%] ···  distance_metric.SEuclideanDistanceBenchmark.time_pairwise                                                                   failed
[ 92.86%] ··· ======= ======== ======== ========
              --                  d             
              ------- --------------------------
                 n       5        10      100   
              ======= ======== ======== ========
                100    failed   failed   failed 
                1000   failed   failed   failed 
               10000   failed   failed   failed 
              ======= ======== ======== ========

[100.00%] ···  distance_metric.WMinkowskiDistanceBenchmark.time_pairwise                                                               1/9 failed
[100.00%] ··· ======= ============= ============= =============
              --                          d                    
              ------- -----------------------------------------
                 n          5             10           100     
              ======= ============= ============= =============
                100    1.60±0.01ms   2.78±0.01ms   24.7±0.07ms 
                1000     159±1ms       278±2ms      2.43±0.01s 
               10000    16.3±0.01s    28.1±0.04s      failed   
              ======= ============= ============= =============

       before           after         ratio
     [b5e5db4a]       [47438839]
     <main>           <remove-old-np-compat-fixtures>
+         864±2μs      1.35±0.01ms     1.56   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(100, 10)
+      86.4±0.6ms        134±0.7ms     1.55   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(1000, 10)
+      8.99±0.01s       13.9±0.01s     1.54   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(10000, 10)
+      28.5±0.1ms       42.4±0.2ms     1.49   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(1000, 5)
+       291±0.8μs        431±0.9μs     1.48   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(100, 5)
+      3.21±0.01s       4.57±0.01s     1.42   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(10000, 5)
+      8.33±0.02s       10.9±0.03s     1.31   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(1000, 100)
+      85.2±0.1ms        112±0.3ms     1.31   distance_metric.MahalanobisDistanceBenchmark.time_pairwise(100, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.


</pre>
</details>

Reply:
It might be interesting to try to do a minimal reproducer and report it upstream to the Cython develop along with the benchmark script and results.

For the time being, I would put the current PR on hold.

Reply:
Following the cython's issue template, I posted [a conversation](https://groups.google.com/g/cython-users/c/rU0_mt3Vgn0/m/1DytI0fKAAAJ) on the [cython-users](https://groups.google.com/g/cython-users) as it's more likely a support discussion. Its snippet is available [here](https://gist.github.com/jjerphan/4906c0a0ae35c76e83b409741fa8c13a). 

Initially, there is no interaction with CPython in `rdist`, as it's the case here in [the surrogate function](https://gist.github.com/jjerphan/4906c0a0ae35c76e83b409741fa8c13a#file-mahalanobis-pyx-L70-L90).

Reply:
Does the benchmark improve with the following?

```
    cdef inline DTYPE_t rdist(self, const DTYPE_t* x1, const DTYPE_t* x2,
                              ITYPE_t size) nogil except -1:
        cdef DTYPE_t tmp, d = 0
        cdef np.intp_t i, j
        cdef:
            DTYPE_t[::1] vec = self.vec
            DTYPE_t[:, ::1] mat = self.mat


        # compute (x1 - x2).T * VI * (x1 - x2)
        for i in range(size):
            vec[i] = x1[i] - x2[i]

        for i in range(size):
            tmp = 0
            for j in range(size):
                tmp += mat[i, j] * vec[j]
            d += tmp * vec[i]
        return d
```

![Screen Shot 2021-06-29 at 9 40 00 AM](https://user-images.githubusercontent.com/5402633/123807674-10ab9c00-d8be-11eb-812d-99c7b0c898b0.png)

Reply:
This seems that it makes it slower.
```python
In [5]: %timeit original_distance.pairwise(X, Y)
543 µs ± 3.96 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [6]: %timeit current_distance.pairwise(X, Y)
1.04 ms ± 29.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: %timeit local_cdef_distance.pairwise(X, Y)
1.28 ms ± 6.05 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

From what I understand from the generated code, is that a check might be (always?) needed when working with memory-views used as attributes as those can be uninitialized. This would also explain the interaction with CPython introduced when calling `rdist`, in the current (as of 4743883) and this last proposed implementation.

---

<details>
<summary> Full snippet </summary>

```cython
%%cython --annotate
#cython: boundscheck=False
#cython: wraparound=False
#cython: cdivision=True

import numpy as np
cimport numpy as np
np.import_array()

ctypedef np.float64_t DTYPE_t
ctypedef np.intp_t ITYPE_t

from libc.math cimport sqrt

DTYPE = np.float64
ITYPE = np.intp

######################################################################
# Numpy 1.3-1.4 compatibility utilities
cdef DTYPE_t* get_vec_ptr(np.ndarray[DTYPE_t, ndim=1, mode='c'] vec):
    return &vec[0]

cdef DTYPE_t* get_mat_ptr(np.ndarray[DTYPE_t, ndim=2, mode='c'] mat):
    return &mat[0, 0]
######################################################################

cdef class Distance():
    
    def pairwise(self, X, Y=None):
        cdef np.ndarray[DTYPE_t, ndim=2, mode='c'] Xarr
        cdef np.ndarray[DTYPE_t, ndim=2, mode='c'] Yarr
        cdef np.ndarray[DTYPE_t, ndim=2, mode='c'] Darr

        Xarr = np.asarray(X, dtype=DTYPE, order='C')
        if Y is None:
            Darr = np.zeros((Xarr.shape[0], Xarr.shape[0]),
                         dtype=DTYPE, order='C')
            self.pdist(Xarr, Darr)
        else:
            Yarr = np.asarray(Y, dtype=DTYPE, order='C')
            Darr = np.zeros((Xarr.shape[0], Yarr.shape[0]),
                         dtype=DTYPE, order='C')
            self.cdist(Xarr, Yarr, Darr)
        return Darr
    
    
    cdef DTYPE_t rdist(self, const DTYPE_t* x1, const DTYPE_t* x2,
                       ITYPE_t size) nogil except -1:
        return -999

    cdef inline DTYPE_t dist(self, const DTYPE_t* x1, const DTYPE_t* x2,
                             ITYPE_t size) nogil except -1:
        return sqrt(self.rdist(x1, x2, size))
    
    cdef int pdist(self, const DTYPE_t[:, ::1] X, DTYPE_t[:, ::1] D) except -1:
        """compute the pairwise distances between points in X"""
        cdef ITYPE_t i1, i2
        for i1 in range(X.shape[0]):
            for i2 in range(i1, X.shape[0]):
                D[i1, i2] = self.dist(&X[i1, 0], &X[i2, 0], X.shape[1])
                D[i2, i1] = D[i1, i2]
        return 0

    cdef int cdist(self, const DTYPE_t[:, ::1] X, const DTYPE_t[:, ::1] Y,
                   DTYPE_t[:, ::1] D) except -1:
        """compute the cross-pairwise distances between arrays X and Y"""
        cdef ITYPE_t i1, i2
        if X.shape[1] != Y.shape[1]:
            raise ValueError('X and Y must have the same second dimension')
        for i1 in range(X.shape[0]):
            for i2 in range(Y.shape[0]):
                D[i1, i2] = self.dist(&X[i1, 0], &Y[i2, 0], X.shape[1])
        return 0
    

cdef class LocalCDEFMahalanobisDistance(Distance):    
    cdef DTYPE_t p
    cdef DTYPE_t[::1] vec
    cdef DTYPE_t[:, ::1] mat
    cdef ITYPE_t size
    cdef object func
    cdef object kwargs
    
    def __cinit__(self):
        self.p = 2
        self.vec = np.zeros(1, dtype=DTYPE, order='c')
        self.mat = np.zeros((1, 1), dtype=DTYPE, order='c')
        self.size = 1
    
    def __init__(self, V=None, VI=None):
        if VI is None:
            if V is None:
                raise ValueError("Must provide either V or VI "
                                 "for Mahalanobis distance")
            VI = np.linalg.inv(V)
        if VI.ndim != 2 or VI.shape[0] != VI.shape[1]:
            raise ValueError("V/VI must be square")

        self.mat = np.asarray(VI, dtype=float, order='C')
        self.size = self.mat.shape[0]
        # we need vec as a work buffer
        self.vec = np.zeros(self.size, dtype=DTYPE)


    cdef inline DTYPE_t rdist(self, const DTYPE_t* x1,
                              const DTYPE_t* x2,
                              ITYPE_t size) nogil except -1:
        cdef:
            DTYPE_t tmp, d = 0
            np.intp_t i, j
            DTYPE_t[:, ::1] mat = self.mat
            DTYPE_t[::1] vec = self.vec


        # compute (x1 - x2).T * VI * (x1 - x2)
        for i in range(size):
            vec[i] = x1[i] - x2[i]

        for i in range(size):
            tmp = 0
            for j in range(size):
                tmp += mat[i, j] * vec[j]
            d += tmp * vec[i]
        return d
    

cdef class CurrentMahalanobisDistance(Distance):    
    cdef DTYPE_t p
    cdef DTYPE_t[::1] vec
    cdef DTYPE_t[:, ::1] mat
    cdef ITYPE_t size
    cdef object func
    cdef object kwargs
    
    def __cinit__(self):
        self.p = 2
        self.vec = np.zeros(1, dtype=DTYPE, order='c')
        self.mat = np.zeros((1, 1), dtype=DTYPE, order='c')
        self.size = 1
    
    def __init__(self, V=None, VI=None):
        if VI is None:
            if V is None:
                raise ValueError("Must provide either V or VI "
                                 "for Mahalanobis distance")
            VI = np.linalg.inv(V)
        if VI.ndim != 2 or VI.shape[0] != VI.shape[1]:
            raise ValueError("V/VI must be square")

        self.mat = np.asarray(VI, dtype=float, order='C')
        self.size = self.mat.shape[0]
        # we need vec as a work buffer
        self.vec = np.zeros(self.size, dtype=DTYPE)


    cdef inline DTYPE_t rdist(self, const DTYPE_t* x1,
                              const DTYPE_t* x2,
                              ITYPE_t size) nogil except -1:
        cdef:
            DTYPE_t tmp, d = 0
            np.intp_t i, j

        # compute (x1 - x2).T * VI * (x1 - x2)
        for i in range(size):
            self.vec[i] = x1[i] - x2[i]

        for i in range(size):
            tmp = 0
            for j in range(size):
                tmp += self.mat[i, j] * self.vec[j]
            d += tmp * self.vec[i]
        return d

    
cdef class OriginalMahalanobisDistance(Distance):    
    cdef DTYPE_t p
    cdef np.ndarray vec
    cdef np.ndarray mat
    cdef DTYPE_t* vec_ptr
    cdef DTYPE_t* mat_ptr
    cdef ITYPE_t size
    cdef object func
    cdef object kwargs
    
    def __cinit__(self):
        self.p = 2
        self.vec = np.zeros(1, dtype=DTYPE, order='c')
        self.mat = np.zeros((1, 1), dtype=DTYPE, order='c')
        self.vec_ptr = get_vec_ptr(self.vec)
        self.mat_ptr = get_mat_ptr(self.mat)
        self.size = 1
    
    def __init__(self, V=None, VI=None):
        if VI is None:
            if V is None:
                raise ValueError("Must provide either V or VI "
                                 "for Mahalanobis distance")
            VI = np.linalg.inv(V)
        if VI.ndim != 2 or VI.shape[0] != VI.shape[1]:
            raise ValueError("V/VI must be square")

        self.mat = np.asarray(VI, dtype=float, order='C')
        self.mat_ptr = get_mat_ptr(self.mat)
        self.size = self.mat.shape[0]
        # we need vec as a work buffer
        self.vec = np.zeros(self.size, dtype=DTYPE)
        self.vec_ptr = get_vec_ptr(self.vec)


    cdef inline DTYPE_t rdist(self, const DTYPE_t* x1, const DTYPE_t* x2,
                              ITYPE_t size) nogil except -1:
        cdef DTYPE_t tmp, d = 0
        cdef np.intp_t i, j

        # compute (x1 - x2).T * VI * (x1 - x2)
        for i in range(size):
            self.vec_ptr[i] = x1[i] - x2[i]

        for i in range(size):
            tmp = 0
            for j in range(size):
                tmp += self.mat_ptr[i * size + j] * self.vec_ptr[j]
            d += tmp * self.vec_ptr[i]
        return d
```
```python
n = 100
d = 10

rng = np.random.RandomState(0)
X = rng.random_sample((n, d))
Y = rng.random_sample((n, d))

V = rng.random_sample((d, d))

original_distance = OriginalMahalanobisDistance(V=V)
current_distance = CurrentMahalanobisDistance(V=V)
local_cdef_distance = LocalCDEFMahalanobisDistance(V=V)
```

</details>

Reply:
> From what I understand from the generated code, is that a check might be (always?) needed when working with memory-views used as attributes as those can be uninitialized.

Hmm, maybe the [compiler directive](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html?highlight=boundscheck#compiler-directives): called `initializedcheck`, which can turn this check off would work?

Reply:
It helps for the current version, still the last version is slower.
```python
%timeit original_distance.pairwise(X, Y)
602 µs ± 61.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
%timeit current_distance.pairwise(X, Y)
661 µs ± 3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
%timeit local_cdef_distance.pairwise(X, Y)
1.34 ms ± 6.75 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```